### PR TITLE
Set up AutotriggerAction

### PR DESCRIPTION
**Pull request by alexfish8**
- Thoughts on naming this `AutotriggerAction` vs `AutotriggerHandler`?

### CODE REVIEW

File: async/auto_action.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+import (
+	"sync"
+	"time"
+
+	"github.com/blend/go-sdk/exception"
+)
+
+// NewAutoAction returns a new NewAutoAction
+func NewAutoAction(interval time.Duration) *AutoAction {
```
Comment (wcharczuk) [inline async/auto_action.go]: wouldn't you also need the max count as a required input?

File: async/auto_action.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+				a.Trigger()
+			}
+			a.latch.Stopped()
+			return
+		}
+	}
+}
+
+// Set sets the value
+func (a *AutoAction) SetValue(value interface{}) {
```
Comment (wcharczuk) [inline async/auto_action.go]: i would refactor this to be `Increment` and have it `atomic.AddInt32(&a.counter, 1)`

Comment (wcharczuk) [reply async/auto_action.go]: basically it avoids grabbing the lock to increment, but would want to grab the lock to force the trigger. you can use `atomic.Load` to do the test if after the increment we'd need to trigger (before doing the normal add) there are some patterns in the stdlib and online for how to do increments like this

File: async/auto_action.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+func (a *AutoAction) TriggerAsync() {
+	a.valueLock.Lock()
+	defer a.valueLock.Unlock()
+	a.triggerUnsafeAsync()
+}
+
+// triggerUnsafe calls the handler, if one is set, without acquiring any locks
+func (a *AutoAction) triggerUnsafe() {
+	if a.handler != nil {
+		a.handler(a.value)
```
Comment (wcharczuk) [inline async/auto_action.go]: you can really just have the `handler` or `listener` be a closure and have it implicit it manages a value.

```
var foo string
auto := new(AutoAction)
auto.Handler = func() {
    ext.Send(&foo)
})
foo = "bar"
auto.Handler() // sends "bar"
```
etc.

File: async/auto_action.go [hunk 4]
```diff
@@ -9,12 +9,11 @@ import (
...
 
 // NewAutoAction returns a new NewAutoAction
-func NewAutoAction(interval time.Duration, maxCount int32) *AutoAction {
+func NewAutoAction(interval time.Duration, maxCounter int32) *AutoAction {
 	return &AutoAction{
-		value:          nil,
-		valueLock:      &sync.Mutex{},
-		maxCount:       maxCount,
-		handler:        nil,
+		Mutex:          sync.Mutex{},
```
Comment (wcharczuk) [inline async/auto_action.go]: you don't need to do this specifically; if you embed the mutex it's implicit

Comment (alexfish8) [reply async/auto_action.go]: ah, right

**Pull request merged**